lib/Conversion/RlweToOpenfhe/RlweToOpenfhe.h

```diff
@@ -0,0 +1,259 @@
+  }
+};
+
+template <typename BinOp, typename OpenfheBinOp>
+struct ConvertRlweCiphertextPlaintextOp : public OpConversionPattern<BinOp> {
+  using OpConversionPattern<BinOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      BinOp op, typename BinOp::Adaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    FailureOr<Value> result = getContextualCryptoContext(op.getOperation());
+    if (failed(result)) return result;
+
+    Value cryptoContext = result.value();
+    rewriter.replaceOpWithNewOp<OpenfheBinOp>(
+        op, op.getOutput().getType(), cryptoContext,
+        adaptor.getCiphertextInput(), adaptor.getPlaintextInput());
+    return success();
+  }
+};
+
+template <typename RlweRotateOp>
+struct ConvertRlweRotateOp : public OpConversionPattern<RlweRotateOp> {
+  ConvertRlweRotateOp(mlir::MLIRContext *context)
+      : OpConversionPattern<RlweRotateOp>(context) {}
+
+  using OpConversionPattern<RlweRotateOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      RlweRotateOp op, typename RlweRotateOp::Adaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    FailureOr<Value> result = getContextualCryptoContext(op.getOperation());
+    if (failed(result)) return result;
+
+    Value cryptoContext = result.value();
+    rewriter.replaceOp(op, rewriter.create<openfhe::RotOp>(
+                               op.getLoc(), cryptoContext, adaptor.getInput(),
+                               adaptor.getOffset()));
+    return success();
+  }
+};
+
+inline bool checkRelinToBasis(llvm::ArrayRef<int> toBasis) {
+  if (toBasis.size() != 2) return false;
+  return toBasis[0] == 0 && toBasis[1] == 1;
+}
+
+template <typename RlweRelinOp>
+struct ConvertRlweRelinOp : public OpConversionPattern<RlweRelinOp> {
+  ConvertRlweRelinOp(mlir::MLIRContext *context)
+      : OpConversionPattern<RlweRelinOp>(context) {}
+
+  using OpConversionPattern<RlweRelinOp>::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      RlweRelinOp op, typename RlweRelinOp::Adaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    FailureOr<Value> result = getContextualCryptoContext(op.getOperation());
+    if (failed(result)) return result;
+
+    auto toBasis = adaptor.getToBasis();
+
+    // Since the `Relinearize()` function in OpenFHE relinearizes a ciphertext
+    // to the lowest level (for (1,s)), the `to_basis` of `CKKS.RelinOp` must be
+    // [0,1].
+    if (!checkRelinToBasis(toBasis)) {
+      op.emitError() << "toBasis must be [0, 1], got [" << toBasis << "]";
+      return failure();
+    }
+
+    Value cryptoContext = result.value();
+    rewriter.replaceOpWithNewOp<openfhe::RelinOp>(
+        op, op.getOutput().getType(), cryptoContext, adaptor.getInput());
+    return success();
+  }
+};
+
+// Rewrite extract as a multiplication by a one-hot plaintext, followed by a
+// rotate.
+template <typename RlweExtractOp, typename RlweMulPlainOp,
```

Comment:
I know it's just a refactor (and I know I wrote it originally), but this pattern seems like it could go in a more general place, since it is not actually emitting OpenFHE code. Maybe put the pattern in a RLWEPatterns.h/.cpp for now? Could also go in its own pass.

Reply:
oh yeah, true! yeah, i wouldn't mind putting the pattern in LWE/Dialect/IR/LWEPatterns - in theory this can also go in a number of other passes